tests/test_mpscircuit.py

```diff
@@ -215,20 +250,96 @@ def do_test_proj(test_circucits, external_wavefunction):
+        tensors=mps.get_tensors(),
+        center_position=mps.get_center_position(),
+    )
+    for t1, t2 in zip(newmps.get_tensors(), mps.get_tensors()):
+        np.testing.assert_allclose(
+            tc.backend.numpy(t1), tc.backend.numpy(t2), atol=1e-12
+        )
+
+
+def do_test_measure(test_circuits: type_test_circuits):
+    (
+        c,
+        w_c,
+        mps,
+        w_mps,
+        mps_exact,
+        w_mps_exact,
+    ) = test_circuits
+    index = [6, 5, 2, 9]
+    status = tc.backend.convert_to_tensor([0.1, 0.3, 0.5, 0.7])
+    result_c = c.measure(*index, with_prob=True, status=status)
+    result_mps = mps.measure(*index, with_prob=True, status=status)
+    result_mps_exact = mps_exact.measure(*index, with_prob=True, status=status)
+    np.testing.assert_allclose(result_mps[0], result_c[0], atol=1e-8)
+    np.testing.assert_allclose(result_mps_exact[0], result_c[0], atol=1e-8)
+    np.testing.assert_allclose(result_mps[1], result_c[1], atol=1e-4)
+    np.testing.assert_allclose(result_mps_exact[1], result_c[1], atol=1e-8)
+
+
+def test_MPO_conversion(highp, tfb):
+    O3 = tc.backend.convert_to_tensor(
+        reproducible_unitary(8).reshape((2, 2, 2, 2, 2, 2))
+    )
+    I = tc.backend.convert_to_tensor(np.eye(2).astype("complex128"))
+    gate = tc.gates.Gate(O3)
+
+    MPO3, _ = tc.MPSCircuit.gate_to_MPO(gate, 2, 3, 4)
+    tensor3 = tc.MPSCircuit.MPO_to_gate(MPO3).tensor
+    tensor3 = tc.backend.numpy(tensor3)
+    tensor3_ref = tc.backend.numpy(O3)
+    np.testing.assert_allclose(tensor3, tensor3_ref, atol=1e-12)
+
+    MPO4, _ = tc.MPSCircuit.gate_to_MPO(gate, 1, 3, 4)
+    tensor4 = tc.MPSCircuit.MPO_to_gate(MPO4).tensor
+    tensor4_ref = tc.backend.einsum("ijkabc,pq->ipjkaqbc", O3, I)
+    tensor4 = tc.backend.numpy(tensor4)
+    tensor4_ref = tc.backend.numpy(tensor4_ref)
+    np.testing.assert_allclose(tensor4, tensor4_ref, atol=1e-12)
+
+
 @pytest.mark.parametrize(
     "backend, dtype", [(lf("tfb"), lf("highp")), (lf("jaxb"), lf("highp"))]
 )
 def test_circuits_1(backend, dtype):
+    import time
+
+    begin = time.time()
     circuits = get_test_circuits(False)
+    print("time", time.time() - begin)
+    do_test_canonical(circuits)
     do_test_wavefunction(circuits)
-    do_test_truncation(circuits, 0.9998648317622654, 0.9999264292512574)
+    do_test_truncation(circuits, 0.9987293417932497, 0.999440840610151)
     do_test_amplitude(circuits)
     do_test_expectation(circuits)
     external = external_wavefunction()
-    do_test_fromwavefunction(external)
+    do_test_fromwavefunction(external, 0.1185)
     do_test_proj(circuits, external)
+    do_test_tensor_input(circuits)
+    do_test_measure(circuits)
+    print("time", time.time() - begin)
 
 
 def test_circuits_2(highp):
     circuits = get_test_circuits(True)
-    do_test_truncation(circuits, 0.9705050538783289, 0.984959108658121)
+    do_test_truncation(circuits, 0.9401410770899974, 0.9654331011546374)
+
+
+if __name__ == "__main__":
```

Comment:
why we keep this part?

Reply:
Sorry forgot to delete... was using it to debug